### DISCUSSION TITLE
Need help understanding a piece of Extrude code.

### DISCUSSION THREAD

**Discussion by briansturgill** [elalish/manifold#734] [Q&A] [answered]
I have two questions... 
1) What is (in manifold code) a Genus?
2) What is this piece of code trying to accomplish?
It really need to know this as I need to generate a "zero" cap in the extruder I've been working on.
 
```C++
  if (isCone)
    for (int j = 0; j < polygons.size(); ++j)  // Duplicate vertex for Genus
      vertPos.push_back({0.0f, 0.0f, height});
```
The code appears to be adding a full set (i.e. the size of the incoming CS) of vertices that are all zero points at height.
It seems to be a way to trick the earlier added point indexes into pointing to the same point.
You went to pains earlier to keep from making triangles with zero length sides... at least I think that was the intention...
I'm not sure what this has to do with Genus.
Can I regard it as a magic formula for making a zero point?
I need to do this for the bottom point as well. (One would assume at height zero there.)

Essentially you seem to be making a set of triangles that share a common point.
Other extruder code I've looked at generally did this as an alternate end_cap.
I think it may be a side effect of your need to keep indices and vertices in sync.

To make it more clear, is the triangles pointing to different vertices  that are in fact the same
point necessary? As I add things differently I could just have the generated triangles pointing to the
same vertex index.  Or is there something in Manifold that wants the repeated vertices?
My gut tells me you might have some topological reason for wanting that.

My extruder, alas, is a bit too complex to get the index / vertices right.
I  write triples of vertices that then looks up the proper index in an unordered map.
I would go mad if I tried to keep them in sync as various 2d shapes have to be access both
forward and reverse.
(Yes, adds about 30% to the time, but it's worth it!)

**Comment by elalish** [comment, marked as answer]
It's only foring over the number of polygons, not polygon verts. If you extrude a donut into a cone, you have two polygons, but if you only add one top vert, it's a non-manifold vert otherwise known as a "kissing vert" - imagine a bowtie. Genus is the standard topological [definition](https://en.wikipedia.org/wiki/Genus_(mathematics)), coming from the Euler characteristic, but that assumes there aren't any non-manifold verts. 

The fundamental difference between this Manifold library and many other geometry libraries that have come before is that it allows duplicate verts as necessary to keep the topology sane. Anything that combines all verts that have the same position in space will not be able to retain manifoldness in general. So it basically all comes down to bookkeeping indices.

**Comment by briansturgill** [comment]
@elalish Wow, thanks!
